sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableSpec.scala

```diff
@@ -46,20 +47,55 @@ object ResolveTableSpec extends Rule[LogicalPlan] {
 
     preparedPlan.resolveOperatorsWithPruning(_.containsAnyPattern(COMMAND), ruleId) {
       case t: CreateTable =>
-        resolveTableSpec(t, t.tableSpec, s => t.copy(tableSpec = s))
+        resolveTableSpec(t, t.tableSpec,
+          fakeRelationFromColumns(t.columns), s => t.copy(tableSpec = s))
       case t: CreateTableAsSelect =>
-        resolveTableSpec(t, t.tableSpec, s => t.copy(tableSpec = s))
+        resolveTableSpec(t, t.tableSpec, None, s => t.copy(tableSpec = s))
       case t: ReplaceTable =>
-        resolveTableSpec(t, t.tableSpec, s => t.copy(tableSpec = s))
+        resolveTableSpec(t, t.tableSpec,
+          fakeRelationFromColumns(t.columns), s => t.copy(tableSpec = s))
       case t: ReplaceTableAsSelect =>
-        resolveTableSpec(t, t.tableSpec, s => t.copy(tableSpec = s))
+        resolveTableSpec(t, t.tableSpec, None, s => t.copy(tableSpec = s))
     }
   }
 
+  private def fakeRelationFromColumns(columns: Seq[ColumnDefinition]): Option[LogicalPlan] = {
+    val attributeList = columns.map { col =>
+      AttributeReference(col.name, col.dataType)()
+    }
+    Some(LocalRelation(attributeList))
+  }
+
+  private def analyzeConstraints(
```

Comment:
Is there any other way to do this? Can we restructure the plan so that the analyzer naturally resolves these expressions? I like that we pivoted to `DefaultValueExpression` for default values, rather than using a custom analyzer.

Reply:
I can't think of any other way.
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumnsUtil.scala#L335
This is similar to what we did for column default.

Reply:
Let me think a bit.

Reply:
Can we do something similar to what @cloud-fan did for `OverwriteByExpression` in SPARK-33412?

Reply:
My worry is that we added `DefaultValueExpression` to eventually get rid of the custom analyzer and optimizer for default values. It would be great not to add more dependencies on it.

Reply:
There are some differences here. In a create table statement:
* column default value and option **CANNOT** reference columns
* constraint **CAN** reference columns

Using a default analyzer with a dummy relation is simple, and it can include all other analysis batches other than the main `Resolution` batch.